internal/builder/builder.go

```diff
@@ -0,0 +1,425 @@
+	if err != nil {
+		return fmt.Errorf("failed to open source root '%s': %w", srcDir, err)
+	}
+	defer srcRoot.Close()
+
+	stagingRoot, err := os.OpenRoot(stagingDir)
+	if err != nil {
+		return fmt.Errorf("failed to open staging root '%s': %w", stagingDir, err)
+	}
+	defer stagingRoot.Close()
+
+	// First, analyze the source pattern to understand the copy intent
+	isGlobPattern := containsGlobChars(srcPattern)
+	destEndsWithSlash := strings.HasSuffix(op.To, "/")
+
+	if !isGlobPattern {
+		// Direct path reference - check what it actually is first (cp-like behavior)
+		return applySingleSourceCopy(ctx, srcRoot, srcPattern, stagingRoot, destRelPath, destEndsWithSlash)
+	}
+
+	// Glob pattern - find all matches and copy each
+	matches, err := fs.Glob(srcRoot.FS(), srcPattern)
+	if err != nil {
+		return fmt.Errorf("invalid glob pattern '%s': %w", srcPattern, err)
+	}
+
+	if len(matches) == 0 {
+		return fmt.Errorf("no files match pattern '%s' in source '%s'", srcPattern, srcAlias)
+	}
+
+	// For glob patterns, destination should be a directory (like cp *.txt dest/)
+	for _, match := range matches {
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+
+		// Calculate destination path based on glob pattern type
+		destFile := calculateGlobDestination(srcPattern, match, destRelPath)
+		if err := copyFileWithRoots(ctx, srcRoot, match, stagingRoot, destFile); err != nil {
+			return fmt.Errorf("failed to copy file '%s' to '%s': %w", match, destFile, err)
+		}
+	}
+
+	return nil
+}
+
+// applySingleSourceCopy handles copying a single, non-glob source (file or directory)
+// using cp-like semantics based on the source type and destination format.
+func applySingleSourceCopy(ctx context.Context, srcRoot *os.Root, srcPath string, stagingRoot *os.Root, destPath string, destEndsWithSlash bool) error {
+	// Stat the source first to determine its type (cp-like approach)
+	srcInfo, err := srcRoot.Stat(srcPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return fmt.Errorf("source '%s' does not exist", srcPath)
+		}
+		return fmt.Errorf("failed to stat source '%s': %w", srcPath, err)
+	}
+
+	if srcInfo.IsDir() {
+		return applySingleDirectoryCopy(ctx, srcRoot, srcPath, stagingRoot, destPath, destEndsWithSlash)
+	} else {
+		return applySingleFileCopy(ctx, srcRoot, srcPath, stagingRoot, destPath, destEndsWithSlash)
+	}
+}
+
+// applySingleFileCopy handles copying a single file using cp-like semantics:
+// - file -> dest (no slash) = copy to dest as filename
+// - file -> dest/ (with slash) = copy to dest/filename
+func applySingleFileCopy(ctx context.Context, srcRoot *os.Root, srcPath string, stagingRoot *os.Root, destPath string, destEndsWithSlash bool) error {
+	var finalDestPath string
+
+	if destEndsWithSlash {
+		// Destination is explicitly a directory - use source filename
+		srcFileName := filepath.Base(srcPath)
+		finalDestPath = filepath.Join(destPath, srcFileName)
+	} else {
+		// Destination is the target filename
+		finalDestPath = destPath
```

Comment:
Here, it's possible that a previous copy operation already created `destPath` as a directory. Will `copyFileWithRoots()` error out if that happens? We need to either document that, or test whether `destPath` exists and is a directory to treat it like such

Reply:
I added a test to prove that a file can not override an existing dir.